File: nuido_flow_ai_mini_knowledge/static/src/models/ai/mini_knowledge_tool_node.js
/*!
// THIS FILE IS A PART OF PUBLIC REPOSITORY https://github.com/yonitjio/exploring-odoo
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT
//
// THIS SOFTWARE IS EXPERIMENTAL AND FOR EDUCATIONAL PURPOSE ONLY.
// DO NOT USE IT IN PRODUCTION.
*/
import { NodeModel } from "@nuido/models/node";
import { AiToolPort } from "@nuido_flow_ai/components/ports/ai_tool_port";
export class MiniKnowledgeToolNodeModel extends NodeModel {
    setup() {
        const auxOutId = "aux-out-" + this.id + "-mini-knowledge-tool";
        this.addAuxOutPort(auxOutId, AiToolPort.name, Number.MAX_SAFE_INTEGER, {
            role: "ai-tool"
        });
    }
}

File: nuido_flow_ai_mini_knowledge/flows/ai/mini_knowledge_tool_node.py
# THIS FILE IS A PART OF PUBLIC REPOSITORY https://github.com/yonitjio/exploring-odoo
#
# This software is released under the MIT License.
# https://opensource.org/licenses/MIT
#
# THIS SOFTWARE IS EXPERIMENTAL AND FOR EDUCATIONAL PURPOSE ONLY.
# DO NOT USE IT IN PRODUCTION.
import logging
_logger = logging.getLogger(__name__)

from typing import Annotated, Literal

from odoo.api import Environment

from odoo.addons.nuido_flow.flows.core.base_node import BaseNode
from odoo.addons.nuido_flow_ai.flows.ai.odoo_function_tool import OdooFunctionTool

def get_knowledge_base(id: Annotated[int, "Knowledge base Id."], odoo_env: Environment = None) \
    -> Annotated[str | Literal[False], f"Knowledge base content."]:
    try:
        mk = odoo_env["nuido_flow_ai.mini.knowledge"].browse(id)
        if mk:
            return {
                "content": mk["content"]
            }
        else:
            return False
    except Exception as e:
        _logger.error(f"Failed to retrieve knowledge base: {e}")
        return False

class MiniKnowledgeToolNode(BaseNode):
    def _process(self, params):
        super()._process(params)

        ft = OdooFunctionTool(get_knowledge_base, description="Query knowledge base content by Id.", odoo_env=self.env)

        res = {
            "tools": [ft]
        }

        return res
